subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/TaskReportTask.java

```diff
@@ -56,6 +58,7 @@
 public class TaskReportTask extends ConventionReportTask {
 
     private boolean detail;
+    private final Property<Boolean> showTypes = getProject().getObjects().property(Boolean.class).convention(false);
```

Comment:
It's more idiomatic and it makes a difference for other kinds of properties. In this case it's not as interesting because it's just a flag and it's unlikely that there's going to be any other reason to configure it.

But let's say instead that this was a `DirectoryProperty` and we wanted the convention to be based on an extension's property. If we set the convention in the task, we need to reach into the project and get the extension. This means that you can't use the task unless you have this extension available. This also breaks configuration caching. 

If the property is something simple (a primitive), we still want to try to have a single place for all of the configuration of the task. If the task can also supply configuration, it breaks down when you try to have plugins that build on top of each other. e.g., java-base <- java <- my.java-conventions. Where would the task configuration fit into that mental model?

The way I think about it is that we want to make the tasks as close to just properties + an action and leave all configuration up to plugins.